### PR TITLE
Merge Entry and QueryResult / Add Entry.ofText/ofJson() to replace identity() / Add CentralDogma.push() without Author

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

There are various usability issues with the current common/client API:

- Entry and QueryResult are quite similar to each other with much code duplication.
- Entry.identity() is not useful but cumbersome because it uses `Object` as the type parameter.
- Authentication should determine the author for CentralDogma.push()

Modifications:

- Merge Entry and QueryResult into one concrete implementation
  - Add Entry.revision()
  - Replace all QueryResult usages with Entry
  - Note that Entry is not possible to represent the absence of an
    entry, therefore:
    - The CompletableFuture returned by watchFile() will not complete
      when an entry is removed.
    - Repository.getOrElse() has been replaced with getOrNull().
- Improve user friendliness of Entry
  - Entry.content() is now non-null and will throw IllegalStateException
    if content is unavailable.
  - Add Entry.hasContent() and ifHasContent()
  - Add Entry.contentAsJson() for better user experience
  - Add Entry.contentAsPrettyText()
- Deprecate Query.identity()
  - Add Query.ofText() and ofJson() which are type-safe
  - Deprecate CentralDogma.getFile(..., String path), which was not
    type-safe by its nature
- Deprecate CentralDogma.push() methods which require an Author parameter
  - Add the push() methods without an Author parameter
    - For the legacy client, it defaults to Author.UNKNOWN. For other
      client implementations, it will be set to the currently
      authenticated user.
- Miscellaneous:
  - Narrow down the 'throws' clauses of the Jackson utility methods

Result:

- Less code duplication
- Easier to use Entry API
- Easier to send an identity query
- Deprecations:
  - Query.identity()
  - CentralDogma.getFile(..., String path)
  - CentralDogma.push(..., Author, ...)

### CODE REVIEW

**Comment by trustin** [comment]
Let me try to increase the test coverage tomorrow.

File: server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/RepositoryUtil.java [hunk 1]
```diff
@@ -53,72 +51,42 @@
...
 
-    static <T> CompletableFuture<QueryResult<T>> watch(Repository repo, Revision lastKnownRev, Query<T> query) {
+    static <T> CompletableFuture<Entry<T>> watch(Repository repo, Revision lastKnownRev, Query<T> query) {
 
         requireNonNull(repo, "repo");
         requireNonNull(lastKnownRev, "lastKnownRev");
         requireNonNull(query, "query");
 
         @SuppressWarnings("unchecked")
         final Query<Object> castQuery = (Query<Object>) query;
```
[the remark this answers is not in the document]
Comment (trustin) [reply server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/RepositoryUtil.java]: Fixed :-)

**Comment by trustin** [comment]
Ready for reviews

File: server/src/main/java/com/linecorp/centraldogma/server/internal/admin/service/RepositoryService.java [hunk 2]
```diff
@@ -114,7 +114,7 @@ public RepositoryService(ProjectManager projectManager, CommandExecutor executor
         final Repository repo = projectManager().get(projectName).repos().get(repoName);
         return repo.normalize(new Revision(revision))
                    .thenCompose(normalized -> repo.get(normalized, query))
-                   .thenApply(queryResult -> DtoConverter.convert(path, queryResult));
+                   .thenApply(queryResult -> new EntryWithRevisionDto(DtoConverter.convert(queryResult)));
```
Comment (minwoox) [inline server/src/main/java/com/linecorp/centraldogma/server/internal/admin/service/RepositoryService.java]: Can we just remove `EntryWithREvisionDto` and returns `EntryDto` instead?

File: common/src/main/java/com/linecorp/centraldogma/common/Entry.java [hunk 3]
```diff
@@ -29,98 +38,265 @@
...
     /**
      * Returns the path of this {@link Entry}.
      */
-    String path();
+    public String path() {
+        return path;
+    }
+
+    /**
+     * Returns if this {@link Entry} has content, which is always true if it's not a directory.
```
Comment (minwoox) [inline common/src/main/java/com/linecorp/centraldogma/common/Entry.java]: `{@code true}`?

File: it/src/test/java/com/linecorp/centraldogma/it/updater/CentralDogmaBeanTest.java [hunk 4]
```diff
@@ -77,19 +76,18 @@ public void test() throws Exception {
...
-        client.watchFile("a", "b", Revision.INIT, Query.identity("/a.json"), 5000).join();
+        final Commit rev = client.push("a", "b", Revision.HEAD, "Add c.json",
+                                        Change.ofJsonUpsert("/c.json",
+                                                            '{' +
+                                                            "  \"foo\": 20," +
+                                                            "  \"bar\": \"Y\"," +
+                                                            "  \"qux\": [\"0\", \"1\"]" +
+                                                            '}'))
+                                  .join();
+        System.err.println(rev);
```
[the remark this answers is not in the document]
Comment (trustin) [reply it/src/test/java/com/linecorp/centraldogma/it/updater/CentralDogmaBeanTest.java]: Fixed

File: server/src/main/java/com/linecorp/centraldogma/server/internal/admin/service/RepositoryService.java (hunk 2 above)
Comment (trustin) [reply server/src/main/java/com/linecorp/centraldogma/server/internal/admin/service/RepositoryService.java]: Fixed

File: common/src/main/java/com/linecorp/centraldogma/common/Entry.java (hunk 3 above)
Comment (trustin) [reply common/src/main/java/com/linecorp/centraldogma/common/Entry.java]: Fixed

**Comment by trustin** [comment]
@imasahiro Any comments?

**Comment by trustin** [comment]
Thanks, reviewers :-)

**Comment by imasahiro** [comment]
lgtm

**Pull request merged**